applications/OptimizationApplication/custom_utilities/geometrical/symmetry_utility.cpp

```diff
@@ -0,0 +1,270 @@
+    {
+        NodeTypePointer pnode = *(node_it.base());
+        mListOfNodesInModelPart[counter++] = pnode;
+    }
+
+    const size_t bucket_size = 100;
+    KDTree search_tree(mListOfNodesInModelPart.begin(), mListOfNodesInModelPart.end(), bucket_size);
+
+    if(mAxisSymmetry){
+
+        #pragma omp declare reduction (merge : std::vector<std::pair <NodeTypePointer,NodeVector>> : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()))        
+        std::vector<std::pair <NodeTypePointer,NodeVector>>& rMap = mRotationalSymmetryData.Map;
+        #pragma omp parallel for reduction(merge: rMap)
+        for (auto& r_node : mrModelPart.Nodes()){
+            int num_rotations = (360.0/mRotationalSymmetryData.Angle);
+            NodeVector p_neighbor_rot_nodes;
+            p_neighbor_rot_nodes.resize(num_rotations-1);
+            for(int r_i=1;r_i<num_rotations;r_i++){
+               NodeTypePointer p_rot_node = GetRotatedNode(r_node,r_i*mRotationalSymmetryData.Angle);
+               double distance;
+               NodeTypePointer p_neighbor_rot_node = search_tree.SearchNearestPoint(*p_rot_node, distance);
+               p_neighbor_rot_nodes[r_i-1] = p_neighbor_rot_node;
+            }
+            std::pair <NodeTypePointer,NodeVector> nodes_pair;
+            nodes_pair = std::make_pair(mrModelPart.pGetNode(r_node.Id()),p_neighbor_rot_nodes);
+            rMap.push_back(nodes_pair); 
+        }
+    }
+
+    if(mPlaneSymmetry){
+        #pragma omp declare reduction (merge : std::vector<std::pair <NodeTypePointer,NodeTypePointer>> : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()))        
+        std::vector<std::pair <NodeTypePointer,NodeTypePointer>>& rMap = mPlaneSymmetryData.Map;
+        #pragma omp parallel for reduction(merge: rMap)
+        for (auto& r_node : mrModelPart.Nodes()){
+            NodeTypePointer p_ref_node = GetReflectedNode(r_node);
+            double distance;
+            NodeTypePointer p_neighbor_ref_node = search_tree.SearchNearestPoint(*p_ref_node, distance);
+            std::pair <NodeTypePointer,NodeTypePointer> nodes_pair;
+            nodes_pair = std::make_pair(mrModelPart.pGetNode(r_node.Id()),p_neighbor_ref_node);
+            rMap.push_back(nodes_pair);          
+        }
+    }
+
+    KRATOS_INFO("SymmetryUtility::Initialize: ") << " Finished initialization of symmetry utility "<<mUtilName<<" in " << timer.ElapsedSeconds() << " s." << std::endl;    
+}
+
+SymmetryUtility::NodeTypePointer SymmetryUtility::GetRotatedNode(NodeType& rNode, double angle){
+    NodeTypePointer p_new_node = Kratos::make_intrusive<NodeType>(rNode.Id(), rNode[0], rNode[1], rNode[2]);
+
+    const array_3d v1 = rNode.Coordinates()-mRotationalSymmetryData.Point;
+    Matrix rot_mat;
+    GetRotationMatrix(angle,rot_mat);
+    p_new_node->Coordinates() = prod(rot_mat, v1) + mRotationalSymmetryData.Point;
+
+    return p_new_node;
+}
```

Comment:
Consider using or extending stuff already implemented in kratos. For example, there's already a [class for affine transformations](https://github.com/KratosMultiphysics/Kratos/blob/95c4e3efc14044e9974ab79d211c8ddfc4f17987/applications/MeshMovingApplication/custom_utilities/linear_transform.h) (don't be fooled by it's name, it **is** actually an affine transform and not a linear one, I messed up the naming).

I know it's quite challenging to find anything in kratos if you don't know exactly what you're looking for, but here are a few tips:
- you can match most of the numerous typos with [agrep](https://github.com/Wikinaut/agrep)
- as for the *liberal* use of English in class names, you can either try to think like a spaniard would, or try your luck with distant "synonym"s from [thesaurus](https://www.thesaurus.com/).

Reply:
Thanks again for the hints and suggestions. I am also in favor what existing implementations however the function you are suggesting is implemented in MeshMovingApp. Making dependency to another application for one function is not convincing me. Are you aware of any similar function at the core ?!